Automattic/crowdsignal-forms
# Add autosaving capabilities for the NPS block

This patch adds autosave capabilities for the NPS block (and possibly others in the future). Solves c/wCM0MVuu-tr.

I accomplished this by creating a `withAutosave` higher order component which takes two arguments:

- An `onSave` handler, which should return a promise and accept two arguments:
  1. `attributes` which are the current attributes of the to be saved block.
  2. `setAttributes` in case the attributes need to be updated based on the logic inside the `onSave` handler.
- `watchedAttributes` which is a list of attribute keys to watch that should trigger the autosave behavior.

The onSave handler is debounced (I set it to a 1.5s default) which should give enough time to accumulate rapid changes into a larger request.
When the `onSave` handler fails, the HoC will retry the save operation until `RETRY_THRESHOLD` is met, after which a `saveError` prop will be set to `true` so our block can display an appropriate error message.

The HoC also inject a `forceSave` prop which can be used to allow the user to manually retry the save, or to save a new block to Crowdsignal.com as soon as it's created.

![Screen Shot 2021-02-03 at 5 13 04 PM](https://user-images.githubusercontent.com/8056203/106781583-8120fa80-6649-11eb-9c4a-7aa91124d323.png)
![Screen Shot 2021-02-03 at 6 00 02 PM](https://user-images.githubusercontent.com/8056203/106781817-c2b1a580-6649-11eb-9eb7-7cea5b8eee93.png)

# Testing

- Create a new post and create a new NPS block. This block should appear in your Crowdsignal dashboard immediately, without the need to update any of its properties.
- Try changing the title or any of the questions, after a very short delay the changes should be reflected in your Crowdsignal dashboard without the need for any additional actions.
- Switch your browser to offline mode, and try making some more changes. An error banner should appear above the block after a few seconds.
  * Go back into online mode and click the _save_ button. The block should now be saved to Crowdsignal.com correctly.
  * Go back into online mode and make a change to the title or the block questions, the block should correctly autosave now.
